### PR TITLE
Ai hologram shortcut

### PR DESCRIPTION
**Pull request by Luxizzle**
[FEATURE]

## About the PR 

Shortcuts for the AI holograms!

You can hold your point key (B) and click somewhere to spawn a hologram at that location. You will be shown a circular menu to choose which hologram you want to spawn there.

image

## Why's this needed? 

The current button forces you to move to a location and press a button on the UI to spawn a hologram at your current eye position, this is faster and does not force you into a popup. And shortcuts are always nice.

Also comes with some fixes for multiContext.
- Moved all the actions to add buttons to a hud to a single proc.
- Allowed the AI to view multiContext buttons.
- Fixed up some math in the experimentalcircle layout.
- Added the ability to choose a custom layout on `showContextActions`

## Changelog

```
(u)Luxizzle:
(*)The AI can now hold their point key (B) and click on a location to spawn a hologram.
```

### CODE REVIEW

**Comment by mordent-goonstation** [comment]
Really like the idea, have not had a chance to look at code yet. Developers: if code looks good feel free to merge with my blessing.

**Comment by Kyle2143** [comment]
Yes, this is good. This is what I *wanted* to do for the original but some MORDENT said to just use an input box. Now he changes his tune. Smh

**Comment by Luxizzle** [comment]
Haha, thats great 😄

**Comment by Tarmunora** [comment]
Should have some kind of 12th button to make it symmetrical 😰. Explicit cancel button maybe?

File: code/mob/living/silicon/ai/holograms.dm [hunk 1]
```diff
@@ -1,26 +1,109 @@
...
-		return
-	if (length(holograms)>= max_holograms)
-		boutput(eyecam, "Not enough RAM to project more holograms. Delete others to make room.")
-		return
-	//select from list of icon_state in holograms.dmi
-	var/input = input("Select a type of hologram to display.","Hologram") as null|anything in list("caution","o2","beepsky","up_arrow","down_arrow","left_arrow","right_arrow","happy_face","neutral_face","sad_face","angry_face")
-	if (input)
+/mob/living/silicon/ai
+	contextLayout = new /datum/contextLayout/experimentalcircle(36)
+	var/list/hologramTypes = list(
```
Comment (ZeWaka) [inline code/mob/living/silicon/ai/holograms.dm:4]: why not use children/concrete typesof(/datum/contextAction/ai_hologram) and get the holo_types, you're needlessly repeating

Comment (Luxizzle) [reply code/mob/living/silicon/ai/holograms.dm]: You're right, didn't think of that. Also applied this to `hologramContextActions`.

**Comment by Luxizzle** [comment]
Attempted to use abstracttypesof, but the runtime checker is yelling at me. Not sure how to fix it, so using childrentypesof for now.

**Comment by ZeWaka** [comment]
You were still using the abstract type somewhere

**Comment by Luxizzle** [comment]
It compiles fine locally and selecting AI even when `ABSTRACT_VIOLATION_CRASH` is defined, I'm guessing it gets called somewhere in the runtime checker, but i can't find it

**Comment by Luxizzle** [comment]
Oh, can't use abstract because it makes all the predecessors also abstract and those are used in different actions.

File: code/mob/dead/ai-camera.dm [hunk 2]
```diff
@@ -175,6 +184,9 @@
 			if(src.client.check_key(KEY_SHOCK))
 				src.set_cursor('icons/cursors/shock.dmi')
 				return
+			if(src.client.check_key(KEY_POINT))
```
Comment (ZeWaka) [inline code/mob/dead/ai-camera.dm:187]: why aren't these just else_ifs :flooshed:

File: code/mob/living/silicon/ai.dm [hunk 3]
```diff
@@ -185,6 +187,13 @@ var/list/ai_emotions = list("Happy" = "ai_happy",\
 
 	holograms = list()
 
+	src.hologramContextActions = list()
+	src.hologramTypes = list()
+	for(var/Action in childrentypesof(/datum/contextAction/ai_hologram))
```
Comment (ZeWaka) [inline code/mob/living/silicon/ai.dm:192]: weird caps but ok

File: code/mob/living/silicon/ai/holograms.dm [hunk 4]
```diff
@@ -1,26 +1,96 @@
...
+
+	proc/create_hologram()
+		if (!deployed_to_eyecam)
+			boutput(src, "Deploy to an AI Eye first to create a hologram.")
+			return
+		if (length(holograms)>= max_holograms)
+			boutput(eyecam, "Not enough RAM to project more holograms. Delete others to make room.")
+			return
+		//select from list of icon_state in holograms.dmi
+		var/input = input("Select a type of hologram to display.","Hologram") as null|anything in hologramTypes
```
Comment (ZeWaka) [inline code/mob/living/silicon/ai/holograms.dm:13]: do we still want the input() way in? why not just have it via the point/context way?

Comment (Luxizzle) [reply code/mob/living/silicon/ai/holograms.dm]: Made the button place the context menu on the current ai eye's position instead, removing the input and hologramTypes

**Review by ZeWaka — APPROVED** [review]
very good!

**Comment by Sovexe** [comment]
This PR appears to have broken the spawn menu for ghosts

**Comment by Sovexe** [comment]
Ghost menus should be ok again now

**Pull request merged**